### PR TITLE
`Development`: Add safeguards against split Hazelcast clusters

### PR DESCRIPTION
**Pull request by Hialus**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] ~~I added multiple integration tests (Spring) related to the features (with a high test coverage).~~ Not possible via JUnit
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

Currently when starting Artemis it can happen that the Hazelcast cluster is not created properly and some nodes form separate cluster of varying sizes (1+). There is currently no way to fix this, except to restart the affected nodes. The prevention for this so far was to just start the nodes one-by-one with a delay of 20 seconds. This made deployment tedious. 

### Description

I added 2 changes to address this:
1. I added a scheduled task to regularly check if the service registry contains hosts that are not in the Hazelcast cluster and then add those
2. I added a basic configuration for split brain handling. This can help to resolve if the cluster split during runtime into multiple clusters.

As a small side-quest I also added the possibility for nodes to have different Hazelcast ports between one another. This could theoretically replace the old local dev cluster handling, though I left it in place for now.

#### Scenario 1

Prerequisites:
- 1 Admin account on a multi node test server (ideally TS3)

1. Deploy to the test server
2. Go to the admin health page (via the "Server Administration" dropdown in the navbar)
4. Click on the eye on the right for the "Hazelcast" row
5. If there are 3 members: You are done
6. If not, please wait a minute, refresh the page and start at 3. again
7. It can take several minutes until the cluster heals itself, as Hazelcast only checks for a split brain every 2 minutes

#### Scenario 2

Prerequisites:
- Do this on Staging ICL ideally

1. Deploy to staging primary and directly after (possibly even while it is still running) secondary (it is important to start them in quick succession to break the cluster on startup)
2. Do some stuff that stores stuff the Hazelcast cluster (e.g. continously trigger a lot of builds on an ICL server)
3. In the beginning not all builds will be processed (You can check this in the build queue of the course by contionously reloading to get the states of the different nodes)
4. After the cluster merged (maybe check in the admin area) the builds will be processed and the build queue is the same across instances

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

How the health panel should look like:
image

## Summary by CodeRabbit

- **New Features**
	- Enhanced cluster connectivity checks with scheduled tasks in the system configuration.
	- Implemented split brain protection for improved system resilience.

- **Style**
	- Reformatted code indentation in `BuildJobContainerService` to improve readability.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/config/CacheConfiguration.java [hunk 1]
```diff
@@ -102,6 +107,47 @@ public CacheManager cacheManager(HazelcastInstance hazelcastInstance) {
...
+     * This scheduled task regularly checks if all members of the Hazelcast cluster are connected to each other.
+     * This is one countermeasure to a split cluster.
+     */
+    @Scheduled(fixedRate = 2, initialDelay = 1, timeUnit = TimeUnit.MINUTES)
+    public void connectToAllMembers() {
+        if (registration == null) {
+            return;
+        }
+        String serviceId = registration.getServiceId();
+        var hazelcastInstance = Hazelcast.getHazelcastInstanceByName("Artemis");
```
Comment (Mtze) [inline src/main/java/de/tum/in/www1/artemis/config/CacheConfiguration.java:120]: Is it a good idea to hardcode the Instance name here?

Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/config/CacheConfiguration.java]: This was already done in the `hazelcastInstance` method before, so I just reused it.  Though the correct config value was already present in this class, it just was not used here before 🙃 

Thanks for the hint I now improved this as well.

**Review by krusche — APPROVED** [review]
Code changes look good to me 👍 great work!

**Review by pzdr7 — APPROVED** [review]
Tested on TS3 (cluster was split, but healed after a while) and LocalCI staging (I encountered two build queue states that resolved eventually)

**Review by bensofficial — APPROVED** [review]

**Review by pzdr7 — APPROVED** [review]

**Review by bensofficial — APPROVED** [review]
Reapprove

**Review by krusche — APPROVED** [review]
Code looks good to me 👍

**Pull request merged**